File: EditDistance.java
/**
 * LeetCode 72 Edit Distance
 */
class Solution {
    public int[][] memo;

    public int minDistance(String word1, String word2) {
        int maxLen = Math.max(word1.length(),word2.length());
        memo = new int[maxLen][maxLen];
        return dp(word1,word2,word1.length() - 1,word2.length() - 1);
    }

    public int dp(String word1,String word2,int i , int j) {
        if (i == -1) {
            return j + 1;
        }
        if (j == -1) {
            return i + 1;
        }
        if (memo[i][j] != 0) {
            return memo[i][j];
        }
        if (word1.charAt(i) == word2.charAt(j)) {
            memo[i][j] = dp(word1, word2, i - 1, j - 1);
            return memo[i][j];
        } else {
            memo[i][j] = Math.min(Math.min(dp(word1,word2,i, j - 1) + 1,dp(word1,word2,i - 1,j) + 1),dp(word1, word2, i - 1, j - 1) + 1);
            return memo[i][j];
        }
    }
}

File: EditDistance.go
func minDistance(word1 string, word2 string) int {
	n := len(word1)
	m := len(word2)
	dp := make([][]int,n + 1)
	for i := range dp {
		dp[i] = make([]int,m + 1)
	}

	for i := 0;i <= n;i++ {
		for j := 0;j <= m;j++ {
			if i == 0 {
				dp[i][j] = j
			} else if j == 0 {
				dp[i][j] = i
			} else {
				if word1[i - 1] == word2[j - 1] {
					dp[i][j] = dp[i - 1][j - 1]
				} else {
					dp[i][j] = min(dp[i - 1][j - 1],min(dp[i][j - 1],dp[i - 1][j])) + 1
				}
			}
		}
	}
	return dp[n][m]
}

func min(a int,b int) int {
	if (a < b) {
		return a
	}
	return b
}
